src/utilities/graphql/transform.ts

```diff
@@ -86,131 +85,330 @@ function nullIfDocIsEmpty(doc: DocumentNode) {
+  directives.forEach(directive => {
+    if (directive.name) {
+      nameSet.add(directive.name);
+    } else if (directive.test) {
+      tests.push(directive.test);
+    }
+  });
+
+  return (directive: DirectiveNode) => (
+    nameSet.has(directive.name.value) ||
+    tests.some(test => test(directive))
+  );
+}
+
+// Helper interface and function used by removeDirectivesFromDocument to keep
+// track of variable references and fragments spreads found within a given
+// operation or fragment definition.
+interface InternalInUseInfo {
+  variables: Set<string>;
+  fragmentSpreads: Set<string>;
+  // Set to true when we deliberately remove a fragment definition, so we can
+  // make sure also to remove dangling ...spreads that refer to it.
+  removed?: boolean;
+  // Populated by the populateTransitiveVars helper function below.
+  transitiveVars?: Set<string>;
+}
+function makeInUseGetterFunction<TKey>() {
+  const map = new Map<TKey, InternalInUseInfo>();
+
+  return function getInUse(key: TKey): InternalInUseInfo {
+    let inUse = map.get(key);
+    if (!inUse) {
+      map.set(key, inUse = {
+        // Variable and fragment spread names used directly within this
+        // operation or fragment definition, as identified by key. These sets
+        // will be populated during the first traversal of the document in
+        // removeDirectivesFromDocument below.
+        variables: new Set,
+        fragmentSpreads: new Set,
+      });
+    }
+    return inUse;
   };
 }
 
 export function removeDirectivesFromDocument(
   directives: RemoveDirectiveConfig[],
   doc: DocumentNode,
 ): DocumentNode | null {
-  const variablesInUse: Record<string, boolean> = Object.create(null);
-  let variablesToRemove: RemoveArgumentsConfig[] = [];
+  const getInUseByOperationName = makeInUseGetterFunction<string>();
+  const getInUseByFragmentName = makeInUseGetterFunction<string>();
+  const getInUse = (
+    ancestors: readonly (ASTNode | readonly ASTNode[])[],
+  ): InternalInUseInfo | null => {
+    for (
+      let p = 0, ancestor: ASTNode | readonly ASTNode[];
+      p < ancestors.length && (ancestor = ancestors[p]);
+      ++p
+    ) {
+      if (isArray(ancestor)) continue;
+      if (ancestor.kind === Kind.OPERATION_DEFINITION) {
+        // If an operation is anonymous, we use the empty string as its key.
+        return getInUseByOperationName(ancestor.name ? ancestor.name.value : "");
+      }
+      if (ancestor.kind === Kind.FRAGMENT_DEFINITION) {
+        return getInUseByFragmentName(ancestor.name.value);
+      }
+    }
+    invariant.error(`Could not find operation or fragment`);
+    return null;
+  };
 
-  const fragmentSpreadsInUse: Record<string, boolean> = Object.create(null);
-  let fragmentSpreadsToRemove: RemoveFragmentSpreadConfig[] = [];
+  let operationCount = 0;
+  for (let i = doc.definitions.length - 1; i >= 0; --i) {
+    if (doc.definitions[i].kind === 'OperationDefinition') {
```

Comment:
[nit] Could we consider using `Kind` everywhere so we don't have to know the string mapping? This also lets us avoid the mix of comparing against `Kind` or strings throughout this module.

```suggestion
    if (doc.definitions[i].kind === Kind.OPERATION_DEFINITION) {
```